File: 134.gas-station.cpp
/*
 * @lc app=leetcode id=134 lang=cpp
 *
 * [134] Gas Station
 */
#include <vector>

using namespace std;

// @lc code=start
class Solution {
public:
    /**
     * Core ideas:
     * 1. Greedy: If car starts at A and cannot reach B, then any station between A and B
     *    cannot reach B. B is the first station that A cannot reach.
     * 2. If the total gas is greater than the total cost, there must be a solution.
     */
    // T: O(n)
    // S: O(1)
    int canCompleteCircuit(vector<int>& gas, vector<int>& cost) {
        const int n = gas.size();
        int tot_gas = 0, cur_gas = 0;
        int start = 0;
        for (int i = 0; i < n; ++i) {
            int diff = gas[i] - cost[i];
            tot_gas += diff;
            cur_gas += diff;
            if (cur_gas < 0) {
                // Can't reach this station. Try 
                // starting from the next one.
                start = i + 1;
                cur_gas = 0;
            }
        }
        // Based on the 2nd idea above, we don't need to simulate the circular run with mod etc.
        return tot_gas < 0 ? -1 : start; 
    }
};
// @lc code=end

File: 134.gas-station.go
/*
 * @lc app=leetcode id=134 lang=golang
 *
 * [134] Gas Station
 */

// @lc code=start
// T: O(n)
// S: O(1)
func canCompleteCircuit(gas []int, cost []int) int {
	curGas := 0
	netGas := 0
	start := 0

	for i := range gas {
		gasGain := gas[i] - cost[i]
		curGas += gasGain
		if curGas < 0 {
			// Greedy: When it's impossible to reach the next station from the current start,
			// any stations between the current start and the current station (incl.) cannot
			// reach the next station. We skip these and start at the next station instead.
			curGas = 0
			start = i + 1
		}
		netGas += gasGain
	}

	// There must be a solution if the total gas is greater than the total cost.
	// No need to go through the stations again to simulate circle.
	if netGas >= 0 {
		return start
	}
	return -1
}

// @lc code=end

